core/src/main/java/org/apache/druid/java/util/metrics/AllocationMetricCollector.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.metrics;
+
+import org.apache.druid.java.util.common.logger.Logger;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+class AllocationMetricCollector
+{
+  private static final Logger log = new Logger(AllocationMetricCollector.class);
+  private Method getThreadAllocatedBytes;
+  private ThreadMXBean threadMXBean;
+  private Map<Long, Long> previousResults;
+  private boolean initialized = false;
+
+  AllocationMetricCollector()
+  {
+    try {
+      // no Java official public API is present for the following method, so using reflection
+      threadMXBean = ManagementFactory.getThreadMXBean();
+      getThreadAllocatedBytes = threadMXBean.getClass().getMethod("getThreadAllocatedBytes", long[].class);
+      getThreadAllocatedBytes.setAccessible(true);
+      long[] allThreadIds = threadMXBean.getAllThreadIds();
+      long[] bytes = (long[]) getThreadAllocatedBytes.invoke(threadMXBean, (Object) allThreadIds);
+
+      previousResults = new HashMap<>(allThreadIds.length);
+      for (int i = 0; i < allThreadIds.length; i++) {
+        long threadId = allThreadIds[i];
+        previousResults.put(threadId, bytes[i]);
+
+      }
+      initialized = true;
+    }
+    catch (Exception e) {
+      log.warn(e, "Cannot initialize %s", AllocationMetricCollector.class.getName());
+    }
+  }
+
+  public Optional<Long> calculateDelta()
+  {
+    if (!initialized) {
+      return Optional.empty();
+    }
+    try {
```

Comment:
Also it would be useful to add a comment with the expected (i. e. default) emit period of this metric, and the approximate number of threads in a Druid node, so that readers of this code could quantify the GC and computational overhead of computing this metric.

I understand that the overhead is small, but want to make it more obvious for the readers.

Reply:
According to this guy experiments: http://openjdk.5641.n7.nabble.com/ThreadMXBean-getCurrentThreadAllocatedBytes-td343872.html
The call to `getThreadAllocatedBytes` for 500 threads in the worst case takes:
```
perl -e 'print (500*8000/1000/1000);'
4
```
4 ms
`AllocationMetricCollector` takes linear time to caculate delta, for 500 threads it's negligibl. One can call it as freqeuntly as once a second, which is far too frequent for any known purpose to me.

Reply:
Add this to code to make it evident for code readers